File: haskell/app/abc-001-c.hs
--
-- ABC001-C
--

module Main (
  main
) where

import qualified Data.ByteString.Char8 as BS

pows :: [Int]
pows = [2, 15, 33, 54, 79, 107, 138, 171, 207, 244, 284, 326]

dirCode :: [String]
dirCode = [
  "N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE", "S",
  "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"
  ]

main :: IO ()
main = do
  [i0, i1] <- BS.words <$> BS.getLine
  let
    Just (deg, _) = BS.readInt i0
    Just (dis, _) = BS.readInt i1
    --ws = (dis * 10 + 30) `div` 60
    ws = (dis + 3) `div` 6
    --p = calcPower dis'
    p = calcPower ws 0 pows
  putStrLn (selectDir deg p ++ " " ++ show p)

calcPower :: Int -> Int -> [Int] -> Int
calcPower _ wp [] = wp
calcPower ws wp (x:xs)
  |ws <= x = wp
  |otherwise = calcPower ws (wp+1) xs

selectDir :: Int -> Int -> String
selectDir deg power
  | power == 0 = "C"
  | otherwise  = dirCode !! (((deg + 112) `mod` 3600) `div` 225)

File: rust/src/bin/abc-001-c.rs
// ABC001-C
//

use std::io;

const POWS: [i32; 12] = [2, 15, 33, 54, 79, 107, 138, 171, 207, 244, 284, 326];
const DIRS: [&str; 16] = [
  "N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
  "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"
  ];

fn main() {
  let (deg, dis) = input();
  let (d, p) = logic(deg, dis);
  output(d, p)
}

fn input() -> (i32, i32) {
  let mut line = String::new();
  io::stdin().read_line(&mut line).ok();
  //println!("line={}", line);
  let mut a: Vec<i32> = vec![];
  for v in line.trim().split(" ") {
    //println!("v={}", v);
    a.push(v.parse().unwrap());
  }
  (a[0], a[1])
}

fn output(d: String, p: i32) {
  println!("{} {}", d, p)
}

fn logic(deg: i32, dis: i32) -> (String, i32) {
  let ws = (dis + 3) / 6;
  let p = calc_power(ws);
  let d = select_dir(deg, p);
  (d, p)
}

fn calc_power(ws: i32) -> i32 {
  let mut wp: i32 = 0;
  for p in POWS.iter() {
    if ws <= *p {
      break;
    }
    wp += 1
  }
  wp
}

fn select_dir(deg: i32, p: i32) -> String {
  if p == 0 {
    "C".to_string()
  } else {
    let i: usize = (((deg + 112) % 3600) / 225) as usize;
    DIRS[i].to_string()
  }
}


// ---
